mlir-sycl/lib/Dialect/Analysis/AliasAnalysis.cpp

```diff
@@ -0,0 +1,125 @@
+  return op->getDialect()->getNamespace() ==
+         sycl::SYCLDialect::getDialectNamespace();
+}
+
+// Return true is the given type \p ty is a MemRef type with a SYCL element
+// type.
+static bool isMemRefOfSYCLType(Type ty) {
+  if (auto mt = dyn_cast<MemRefType>(ty))
+    return sycl::isSYCLType(mt.getElementType());
+  return false;
+}
+
+//===----------------------------------------------------------------------===//
+// AliasAnalysis
+//===----------------------------------------------------------------------===//
+
+AliasResult sycl::AliasAnalysis::aliasImpl(Value lhs, Value rhs) {
+  if (lhs == rhs)
+    return AliasResult::MustAlias;
+  if (Optional<AliasResult> aliasResult = handleRestrictAlias(lhs, rhs))
+    return *aliasResult;
+  if (Optional<AliasResult> aliasResult = handleSYCLAlias(lhs, rhs))
+    return *aliasResult;
+
+  return LocalAliasAnalysis::aliasImpl(lhs, rhs);
+}
+
+Optional<AliasResult> sycl::AliasAnalysis::handleRestrictAlias(Value lhs,
+                                                               Value rhs) {
+  // Function arguments do not alias if any of them are 'restrict' qualified.
+  if (isFuncArg(lhs) && isFuncArg(rhs))
+    if (isRestrict(lhs) || isRestrict(rhs))
+      return AliasResult::NoAlias;
+
+  return std::nullopt;
+}
+
+Optional<AliasResult> sycl::AliasAnalysis::handleSYCLAlias(Value lhs,
+                                                           Value rhs) {
+  Operation *lhsOp = lhs.getDefiningOp(), *rhsOp = rhs.getDefiningOp();
+  if (!isSYCLOperation(lhsOp) && !isSYCLOperation(rhsOp))
+    return std::nullopt;
+
+  // Handle accessor.subscript operations.
+  if (auto aliasResult = handleAccessorSubscriptAlias(lhs, rhs))
+    return *aliasResult;
+
+  // TODO: handle the other SYCL operations.
+
+  return std::nullopt;
+}
+
+Optional<AliasResult>
+sycl::AliasAnalysis::handleAccessorSubscriptAlias(Value lhs, Value rhs) {
+  Operation *lhsOp = lhs.getDefiningOp(), *rhsOp = rhs.getDefiningOp();
+  auto lhsSubOp = dyn_cast_or_null<sycl::SYCLAccessorSubscriptOp>(lhsOp);
+  auto rhsSubOp = dyn_cast_or_null<sycl::SYCLAccessorSubscriptOp>(rhsOp);
+
+  auto typesDoNotAlias = [](Type lhsTy, Type rhsTy) {
+    return (lhsTy != rhsTy && isMemRefOfSYCLType(rhsTy));
+  };
+
+  // The value produced by an accessor subscript operation does not alias
+  // with values that have a different SYCL type.
+  // Example:
+  //   %1 = affine.load %alloca[0] : memref<1x!sycl_id_1>
+  //   %2 = sycl.accessor.subscript %arg0[%alloca_0] : memref<?xf32, 4>
+  // here 'alias(%2, %alloca) = NoAlias' because:
+  //   - the types of the value '%alloca' and '%2' types are different, and
+  //   - %2 is the result of an accessor.subscript operation, and
+  //   - %alloca type (memref<1x!sycl_id_1>) is a MemRef of a SYCL type
+  Type lhsTy = lhs.getType(), rhsTy = rhs.getType();
+  if ((lhsSubOp && typesDoNotAlias(lhsTy, rhsTy)) ||
+      (rhsSubOp && typesDoNotAlias(rhsTy, lhsTy)))
+    return AliasResult::NoAlias;
```

Comment:
I don't think this is correct... We might have two accessors pointing to the same region of memory and coming from two different buffers:
```cpp
#include <sycl/sycl.hpp>

#include <vector>
#include <iostream>

using namespace sycl;

int main() {
  queue q;
  std::vector<float> v(10);
  {
    buffer b{v};
    auto clone = b.reinterpret<int>();
    q.submit([&](handler &cgh) {
               accessor acc{b, cgh, write_only, no_init};
               accessor cloneacc{clone, cgh, write_only, no_init};
               cgh.parallel_for(10, [=](id<1> i) {
                                      acc[i] = i[0];
                                      cloneacc[i] = i[0] + 1;
                                    });
             });
  }
  for (auto el : v) {
    std::cout << el << "\n";
  }
}
```
Will print:
```
1.4013e-45
2.8026e-45
4.2039e-45
5.60519e-45
7.00649e-45
8.40779e-45
9.80909e-45
1.12104e-44
1.26117e-44
1.4013e-44
```

Reply:
could still be fine, as `typesDoNotAlias ` checks that one of the type is a memref to a sycl type.

Reply:
What @whitneywhtsang is correct. We are being conservative, so in this example we would return 'MayAlias' because, 
in the scenario above, the 2 accessors subscript operations would return values of type  `memref<?xf32, 4>` and  `memref<?xi32, 4>`, and neither is a SYCL type. 

P.S.: The scenario above is tricky. An optimizing compiler that respects ANSI aliasing rules (TBAA) would consider the 2 types as not being aliased, possibly leading to unexpected behavior. IMO the type punning caused by the use of the reinterpret cast makes the program illegal under strict aliasing rules.

Reply:
> could still be fine, as `typesDoNotAlias ` checks that one of the type is a memref to a sycl type.

Missed that, sorry. But that does not fix the issue:

```cpp
int main() {
  constexpr std::size_t num_ids{3};
  queue q;
  std::vector<float> v(num_ids * sizeof(id<1>) / sizeof(float));
  {
    buffer b{v};
    auto clone = b.reinterpret<id<1>>();
    q.submit([&](handler &cgh) {
               accessor acc{b, cgh, write_only, no_init};
               accessor cloneacc{clone, cgh, write_only, no_init};
               cgh.parallel_for(3, [=](id<1> i) {
                                      acc[i] = i[0];
                                      cloneacc[i] = i;
                                    });
             });
  }
  for (auto el : v) {
    std::cout << el << "\n";
  }
}
```
Still problematic.


> P.S.: The scenario above is tricky. An optimizing compiler that respects ANSI aliasing rules (TBAA) would consider the 2 types as not being aliased, possibly leading to unexpected behavior. IMO the type punning caused by the use of the reinterpret cast makes the program illegal under strict aliasing rules.

This is not a `reinterpret_cast`, but a call to `buffer::reinterpret`. [By the SYCL spec](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:buffers), we have:

> A SYCL buffer can construct an instance of a SYCL buffer that reinterprets the original SYCL buffer with a different type, dimensionality and range using the member function reinterpret. The reinterpreted SYCL buffer that is constructed must behave as though it were a copy of the SYCL buffer that constructed it (see [Section 4.5.2](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:reference-semantics)) with the exception that the type, dimensionality and range of the reinterpreted SYCL buffer must reflect the type, dimensionality and range specified when calling the reinterpret member function. By extension of this, the class member types value_type, reference and const_reference, and the member functions get_range() and size() of the reinterpreted SYCL buffer must reflect the new type, dimensionality and range. The data that the original SYCL buffer and the reinterpreted SYCL buffer manage remains unaffected, though the representation of the data when accessed through the reinterpreted SYCL buffer may alter to reflect the new type, dimensionality and range. It is important to note that a reinterpreted SYCL buffer is a copy of the original SYCL buffer only, and not a new SYCL buffer. Constructing more than one SYCL buffer managing the same host pointer is still undefined behavior.

I don't think we can ensure two `accessor`s do not overlap without host-side analysis. Maybe we can discuss this on Monday in the weekly call anyway.

Reply:
Yes, to deal with programs that use type punning operations we would have to have visibility not just on the host code but, in the worse case scenario, the entire program (because one might cast a pointer to a different pointer type in a function the compiler does not "see" and return that pointer). 

However all such programs violate ANSI aliasing rules (also called TBAA or strict aliasing). Compilers usually deal with such programs by providing a compiler option (e.g -q[no]strict for the IBM XL compiler, -f[no]-strict-aliasing for gcc, ...).  
Under `-fno-strict-aliasing` the program is then allowed to violate ANSI aliasing rules and use type punning operations. 
In the absence of such option the compiler is allwed to assume the program abide to ANSI alasing rules. 

I think clang has a -fno-strict-aliasing (https://llvm.org/devmtg/2017-10/slides/Finkel-The%20Type%20Sanitizer.pdf_) so we can trigger the code under discussion via that option. Lets discuss this on Monday.

Reply:
We decided to handle the -fno-strict-aliasing in a separate PR and merge this in.